### DISCUSSION TITLE
Simple Web Access

### DISCUSSION THREAD

**Discussion by TheCyberViking** [oobabooga/text-generation-webui#932] [Ideas]
add an extension with an on-off switch that allows the prompt or part of it the prompt to be sent to Google and the returned data to be used to generate the response. for example, "write the following into a report <search:> What's the weather in Berlin?" the app sends a query to Google, it takes a response and generates the returned data like "The Weather today in Berlin, Germany is 75f" or something. I attempted to write this myself, but haven't had the time to get it working, maybe someone else can add to it or get it working? 

```
import gradio as gr
import modules.shared as shared
from googlesearch_python.googlesearch import search
import requests
from bs4 import BeautifulSoup
import re

SEARCH_ACCESS = False

def ui():
    global SEARCH_ACCESS
    SEARCH_ACCESS = gr.inputs.Checkbox(label="Enable Google Search", default=False)
    return [SEARCH_ACCESS]

def input_modifier(user_input):
    global SEARCH_ACCESS
    if SEARCH_ACCESS:
        if user_input.lower().startswith("search:"):
            query = user_input[len("search:"):].strip()
            search_results = search(query, num_results=3)
            search_data = []

            for result in search_results:
                try:
                    response = requests.get(result)
                    soup = BeautifulSoup(response.text, 'html.parser')
                    text = soup.get_text(strip=True)
                    text = re.sub(r'\s+', ' ', text)
                    search_data.append(text)
                except Exception as e:
                    pass

            compiled_data = "\n".join(search_data)
            return f"{user_input}\n\n{compiled_data}"
    return user_input

def output_modifier(output):
    return output

def bot_prefix_modifier(prefix):
    return prefix
```

**Comment by St33lMouse** [comment]
Good idea. Suggest allowing choice of search engine.

**Reply by TomLucidor** [reply]
Definitely + Wikipedia or Fandom + Blogs or Aggregators

**Comment by Timmytim1976** [comment]
Hi I have never coded a programme in Python so I gave it to ChatGPT and this is what it gave as an output when I gave it your code an asked it to improve it. The code seems to work for me and it dose more then just give weather reports.

import gradio as gr
import requests
from bs4 import BeautifulSoup
import re

class GoogleSearch:
    def __init__(self, num_results=3):
        self.num_results = num_results

    def search(self, query):
        search_results = []
        try:
            for j in search(query, num_results=self.num_results):
                search_results.append(j)
        except:
            pass

        search_data = []
        for result in search_results:
            try:
                response = requests.get(result)
                soup = BeautifulSoup(response.text, 'html.parser')
                
                # Check for the type of query and extract relevant information
                if "weather" in query:
                    temperature = soup.find("span", attrs={"class": "wob_t"}).text.strip()
                    location = soup.find("div", attrs={"class": "wob_loc mfMhoc"}).text.strip()
                    date = soup.find("div", attrs={"class": "wob_dts"}).text.strip()
                    search_data.append(f"The weather in {location} on {date} is {temperature}")
                elif "price" in query and ("car" in query or "vehicle" in query):
                    price = soup.find("span", attrs={"class": "a-price-whole"}).text.strip()
                    currency = soup.find("span", attrs={"class": "a-price-symbol"}).text.strip()
                    search_data.append(f"The price of the {query} is {currency}{price}")
                elif "moon landing" in query:
                    date = soup.find("div", attrs={"class": "BNeawe iBp4i AP7Wnd"}).text.strip()
                    search_data.append(f"The next planned moon landing is {date}")
                else:
                    text = soup.get_text(strip=True)
                    text = re.sub(r'\s+', ' ', text)
                    search_data.append(text)
            except Exception as e:
                pass

        compiled_data = "\n".join(search_data)
        return compiled_data

SEARCH_ACCESS = False

def ui():
    global SEARCH_ACCESS
    SEARCH_ACCESS = gr.inputs.Checkbox(label="Enable Google Search", default=False)
    return [SEARCH_ACCESS]

def input_modifier(user_input):
    global SEARCH_ACCESS
    if SEARCH_ACCESS:
        if user_input.lower().startswith("search:"):
            query = user_input[len("search:"):].strip()
            search_obj = GoogleSearch(num_results=3)
            search_data = search_obj.search(query)
            return search_data
    return user_input

def output_modifier(output):
    return output

def bot_prefix_modifier(prefix):
    return prefix

interface = gr.Interface(fn=bot_prefix_modifier,
                          inputs=ui(),
                          outputs=gr.outputs.Textbox(),
                          title='Google Search Extension',
                          description='An extension to allow the app to search Google and generate a response based on the returned data')
interface.launch()

**Reply by Timmytim1976** [reply]
Here's an example of how you can modify the input_modifier function to enable an open search field:

python

def input_modifier(user_input):
    global SEARCH_ACCESS
    if SEARCH_ACCESS:
        if user_input.lower().startswith("search:"):
            query = user_input[len("search:"):].strip()
            search_results = search(query, num_results=3)
            search_data = []

            for result in search_results:
                try:
                    response = requests.get(result)
                    soup = BeautifulSoup(response.text, 'html.parser')
                    text = soup.get_text(strip=True)
                    text = re.sub(r'\s+', ' ', text)
                    search_data.append(text)
                except Exception as e:
                    pass

            compiled_data = "\n".join(search_data)
            return f"{query}\n\n{compiled_data}"
        else:
            return shared.generate_text(user_input)
    else:
        return user_input

With this modification, the input_modifier function will check if the input starts with "search:". If it does, it will extract the search query and perform a Google search. If it doesn't start with "search:", the function will call the generate_text function from the shared module, which can be used to generate a response based on the user input.

Note that you'll need to modify the ui function to remove the checkbox for enabling Google search since it's no longer needed. Here's the updated ui function:

python

def ui():
    return gr.inputs.Textbox(label="Type something here")

**Reply by Timmytim1976** [reply]
improved script with open search.
import gradio as gr
import modules.shared as shared
from googlesearch_python.googlesearch import search
import requests
from bs4 import BeautifulSoup
import re

def input_modifier(user_input):
    if user_input.strip().lower().startswith("search:"):
        query = user_input[len("search:"):].strip()
        search_results = search(query, num_results=3)
        search_data = []

        for result in search_results:
            try:
                response = requests.get(result)
                soup = BeautifulSoup(response.text, 'html.parser')
                text = soup.get_text(strip=True)
                text = re.sub(r'\s+', ' ', text)
                search_data.append(text)
            except Exception as e:
                pass

        compiled_data = "\n".join(search_data)
        return f"{query}\n\n{compiled_data}"
    else:
        return shared.generate_text(user_input)

def output_modifier(output):
    return output

def bot_prefix_modifier(prefix):
    return prefix

iface = gr.Interface(
    fn=shared.chat,
    inputs=gr.inputs.Textbox(label="Type something here"),
    outputs=gr.outputs.HTML(),
    input_modifiers=[input_modifier],
    output_modifier=output_modifier,
    title="Chat with ChatGPT",
    description="Talk to a machine learning model trained by OpenAI."
)

if __name__ == "__main__":
    iface.launch()

**Comment by StefanIsMe** [comment]
I would love to contribute to this, but could you please give me some instructions on where and how to install the script?

**Reply by Timmytim1976** [reply]
For me I would just copy it to Notepad++ and then save it as a python file named as "script.py" and place in a new folder in " C:\oobabooga-windows\text-generation-webui\extensions\" I choose to call  it "websearch". Then using "Notepad++" open the "start-webui" and go to line that reads somthing like this "open call python server.py --auto-devices --cai-chat --wbits 4 --groupsize 128" and add this " --extension websearch" to the end of the line and save it. Warring this is not fully tested and is very messy and I am not a programmer.

**Reply by StefanIsMe** [reply]
Thank you for the instructions.

**Comment by StefanIsMe** [comment]
I'm trying to run the original script from TheCyberViking but I always have the following results in cmd

```
Starting the web UI...

===================================BUG REPORT===================================
Welcome to bitsandbytes. For bug reports, please submit your error trace to: https://github.com/TimDettmers/bitsandbytes/issues
================================================================================
CUDA SETUP: CUDA runtime path found: D:\repo\Vicuna\oobabooga-windows\installer_files\env\bin\cudart64_110.dll
CUDA SETUP: Highest compute capability among GPUs detected: 8.6
CUDA SETUP: Detected CUDA version 117
CUDA SETUP: Loading binary D:\repo\Vicuna\oobabooga-windows\installer_files\env\lib\site-packages\bitsandbytes\libbitsandbytes_cuda117.dll...
Loading anon8231489123_vicuna-13b-GPTQ-4bit-128g...
Found the following quantized model: models\anon8231489123_vicuna-13b-GPTQ-4bit-128g\vicuna-13b-4bit-128g.safetensors
Loading model ...
D:\repo\Vicuna\oobabooga-windows\installer_files\env\lib\site-packages\safetensors\torch.py:99: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  with safe_open(filename, framework="pt", device=device) as f:
D:\repo\Vicuna\oobabooga-windows\installer_files\env\lib\site-packages\torch\_utils.py:776: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()
D:\repo\Vicuna\oobabooga-windows\installer_files\env\lib\site-packages\torch\storage.py:899: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  storage = cls(wrap_storage=untyped_storage)
Done.
Loaded the model in 128.55 seconds.
Loading the extension "websearch"... Fail.
Traceback (most recent call last):
  File "D:\repo\Vicuna\oobabooga-windows\text-generation-webui\modules\extensions.py", line 19, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "D:\repo\Vicuna\oobabooga-windows\text-generation-webui\extensions\websearch\script.py", line 3, in <module>
    from googlesearch_python.googlesearch import search
ModuleNotFoundError: No module named 'googlesearch_python'
Loading the extension "gallery"... Ok.
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
```

I have installed googlesearch with pip
`pip install googlesearch-python
`

And checked it was installed
```
C:\Users\stefa>pip show googlesearch-python
Name: googlesearch-python
Version: 1.2.3
Summary: A Python library for scraping the Google search engine.
Home-page: https://github.com/Nv7-GitHub/googlesearch
Author: Nishant Vikramaditya
Author-email: junk4Nv7@gmail.com
License:
Location: C:\Users\stefa\AppData\Local\Programs\Python\Python311\Lib\site-packages
Requires: beautifulsoup4, requests
Required-by:
```

Does anyone know what's wrong?

**Reply by Timmytim1976** [reply]
Check that the package is installed in the correct location. The output of pip show googlesearch-python indicates that the package is installed in C:\Users\stefa\AppData\Local\Programs\Python\Python311\Lib\site-packages, but your program may be looking for it in a different location. You can try adding the path to the package to your system's PYTHONPATH environment variable.

    Check that the package is being imported correctly in your code. Make sure that you are using the correct import statement to import the search function from the googlesearch_python module. Double-check that the spelling and capitalization of the module name is correct.

    Make sure that any dependencies required by the googlesearch_python module are also installed. The error message indicates that the module requires the beautifulsoup4 and requests packages, so make sure these are installed as well.

If none of these solutions work, you may need to provide more information . Please do note TheCyberViking stated in his post " **I attempted to write this myself, but haven't had the time to get it working, maybe someone else can add to it or get it working?**"

**Reply by StefanIsMe** [reply]
I tried all of those suggestions, but no luck. I also tried your script and had the same result - it can't find the module. Is your version of the script working properly for you?

**Reply by Timmytim1976** [reply]
No sorry :(

**Reply by StefanIsMe** [reply]
Oh, wait. What? Even your script isn't working for you? You can't code from scratch with ChatGPT. I just know the basics of development. ChatGPT is great, but it doesn't have the full codebase to come up with new features from scratch. I only offered to help because I thought you had something working.

**Comment by El-Hondo** [comment]
For a start just change your import from 
`from googlesearch_python.googlesearch import search`
to
`from googlesearch import search`
That makes it run and eliminats that `ModuleNotFoundError: No module named 'googlesearch_python`

Still problems while it runs. But one problem after the other.

**Comment by GitHub1712** [comment]
This works for me now. Create a folder search in extensions. Create script.py with the code below. Start the Server with --extensions search
I changed this to duckduckgo because google is blocking after some requests, could be changed to the google api...

```
import requests
from bs4 import BeautifulSoup

max_results = 3
result_max_characters = 2000
def input_modifier(string):
    if string.strip().lower().startswith("search:"):
        searchstring = string[len("search:"):].strip()
        url = f"https://duckduckgo.com/html?q={searchstring}"
        headers = {'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/78.0.3904.97 Safari/537.36'}
        response = requests.get(url, headers=headers)
        soup = BeautifulSoup(response.text, 'html.parser')
        result_urls = []
        for result in soup.select('.result__url'):
            url = result.text.strip()
            if url != '':  # skip any empty urls
                result_urls.append(url)
        texts = string + ' '
        count = 0
        for result_url in result_urls:
            try:
                if count <= max_results:
                    response = requests.get("https://" + result_url, headers=headers)
                    soup = BeautifulSoup(response.content, 'html.parser')
                    text = soup.get_text().strip().replace('\n', ' ').replace('\r', '').replace('\t', '')
                    texts = texts + ' ' + text
                    count += 1
            except:
                continue
        texts = texts[0:result_max_characters]
        print(texts)
        return texts
    else:
        return string


def output_modifier(string):
    return string


def bot_prefix_modifier(string):
    return string
```

**Reply by hwamil** [reply]
The extension is said to be loaded but AI is not searching anything. Any pointers you could give?

**Reply by Gyztor** [reply]
yup, seems to not be able to utilize it, it can search but it can't parse that information.

**Comment by GitHub1712** [comment]
I am not sure, who wrote this script, you, me, chatGPT?
Interesting eachadea_vicuna-13b-1.1 realy tried hard to make me not connecting him to the internet because it is too dangerous.

**Comment by ghost** [comment]
Can someone review and test this code, I used bing chat to enhance it.

```python
import gradio as gr
import modules.shared as shared
from googlesearch import search
import requests
from bs4 import BeautifulSoup
import re

search_access = False # Use lowercase and underscores for variable names

def ui():
    global search_access
    search_access = gr.components.Checkbox(label="Enable Google Search")
    return [search_access]

def input_modifier(user_input):
    global search_access
    if search_access:
        if user_input.lower().startswith("search:"):
            query = user_input[len("search:"):].strip()
            search_results = search(query, num_results=3)
            # Use list comprehension to create a list of texts from the search results
            search_data = [get_text(result) for result in search_results]

            compiled_data = "\n".join(search_data)
            return f"{user_input}\n\n{compiled_data}"
    return user_input

# Define a helper function to get the text from a URL
def get_text(url):
    try:
        response = requests.get(url)
        soup = BeautifulSoup(response.text, 'html.parser')
        text = soup.get_text(strip=True)
        text = re.sub(r'\s+', ' ', text)
        return text
    except requests.exceptions.RequestException as e: # Use requests exceptions to handle errors
        print(e) # Print the error message
        return "" # Return an empty string

def output_modifier(output):
    return output

def bot_prefix_modifier(prefix):
    return prefix

```

**Comment by ghost** [comment]
@TheCyberViking I finally got it working, [here](https://github.com/simbake/web_search) is the code hosted on github. Many thanks to you all.

**Reply by TheCyberViking** [reply]
Awesome cant wait to try it

**Comment by Mahesh-nrg** [comment]
I have made small changes and made this Google search enabled by default, by the wake-up word as search, code is  here:

 or we can use the EdgeGPT extension for web access, this works with Bing, https://github.com/GiusTex/EdgeGPT

import gradio as gr
import modules.shared as shared
from googlesearch import search
from bs4 import BeautifulSoup
import re
import requests

search_access = True

def input_modifier(user_input, state):
    # print(state['context'])
    global search_access
    if search_access:
        if user_input.lower().startswith("search"):
            shared.processing_message = "*Searching online...*"
            try:
                # print(f'{state["context_instruct"]}')
                query = user_input.strip("search").strip()
                search_results = search(query, num_results=1)
                search_data = [(get_text(result), result) for result in search_results]
                # state[
                #    "context_instruct"
                # ] = "Answer the question according to the google search results provided and provide source where necessary"
                state[
                    "context"
                ] = "Summarize the online results and answer the question correctly, provide links where necessary"
                # print(f"online results: {search_data}")
                user_prompt = f"{user_input}\n online results: {search_data}"
                print(f"{user_prompt}")
                return user_prompt
            except Exception as e:
                # print the type and message of the exception
                print(type(e), e)
                state[
                    "context_instruct"
                ] = "Tell the user they are experiencing connection issues"
                return ""

    shared.processing_message = "*Typing...*"
    return user_input


def output_modifier(output):
    return output


def bot_prefix_modifier(prefix):
    return prefix


def print_data(data):
    return data


def get_text(url):
    response = requests.get(url)
    soup = "\n".join(
        [
            p.text
            for p in BeautifulSoup(response.text, "html.parser").find_all(
                ["section", "p"]
            )
        ]
    )
    soup = soup[:2044]
    text = re.sub(r"\s+", " ", soup.strip())
    output = "\n".join([text, f" -Source: {response.url}"])
    return output

**Reply by TomLucidor** [reply]
The repo got archived?

**Reply by Mahesh-nrg** [reply]
Still,active. what is the target to achieve.  Google Search, Integration with petals are completed

**Reply by TomLucidor** [reply]
Can we just dump URLs into this and have them read a specific article/blog instead? Rather than "phoning in"?

**Comment by beansfotos** [comment]
Keep getting the error: 
2023-12-15 14:43:56 ERROR:Failed to load the extension "web_search".
Traceback (most recent call last):
  File "C:\Users\mail\AI\text-generation-webui\modules\extensions.py", line 36, in load_extensions
    exec(f"import extensions.{name}.script")
  File "<string>", line 1, in <module>
  File "C:\Users\mail\AI\text-generation-webui\extensions\web_search\script.py", line 3, in <module>
    from selenium import webdriver
ModuleNotFoundError: No module named 'selenium'


And I've installed it and checked to make sure Selenium is installed, the package is there. 
If anyone can assist, please do.
